packages/ens-utils/src/ethregistrar.ts

```diff
@@ -3,323 +3,351 @@ import {
   MIN_ETH_REGISTRABLE_LABEL_LENGTH,
   ETH_TLD,
   charCount,
-  getDomainLabelFromENSName,
+  buildENSName,
 } from "./ensname";
-import { NFTRef, TokenId, buildNFTRef, buildTokenId } from "./nft";
-import { namehash, labelhash } from "viem/ens";
-import { buildAddress } from "./address";
-import { ChainId, MAINNET } from "./chain";
-import { ContractRef, buildContractRef } from "./contract";
+import { buildContractRef, ContractRef } from "./contract";
 import {
   Duration,
   SECONDS_PER_DAY,
   Timestamp,
   addSeconds,
   buildDuration,
+  buildTimePeriod,
   formatTimestampAsDistanceToNow,
   now,
 } from "./time";
 import {
   Price,
   addPrices,
   approxScalePrice,
+  buildPrice,
   formattedPrice,
-  multiplyPriceByNumber,
   subtractPrices,
 } from "./price";
 import { Currency } from "./currency";
+import {
+  PrimaryRegistrationStatus,
+  RegistrarChargeType,
+  RegistrarUnsupportedNameError,
+  Registration,
+  scaleAnnualPrice,
+  SecondaryRegistrationStatus,
+  RegistrarCharge,
+  RegistrarTemporaryFee,
+  RegistrarBaseFee,
+  RegistrarSpecialNameFee,
+  RegistrationPriceQuote,
+  RegistrarAction,
+  RenewalPriceQuote,
+  OnchainRegistrar,
+} from "./registrar";
+import { MAINNET } from "./chain";
+import { buildAddress } from "./address";
 
-export interface Registrar {
-  contract: ContractRef;
+export class EthRegistrar implements OnchainRegistrar {
+  protected readonly registrarContract: ContractRef;
 
-  getTokenId(name: ENSName, isWrapped: boolean): TokenId;
-  isClaimable(name: ENSName, isWrapped: boolean): boolean;
-}
+  public constructor(registrarContract: ContractRef) {
+    this.registrarContract = registrarContract;
+  }
 
-// known registrars
-export const WRAPPED_MAINNET_ETH_REGISTRAR_CONTRACT = buildContractRef(
-  MAINNET,
-  buildAddress("0xD4416b13d2b3a9aBae7AcD5D6C2BbDBE25686401"),
-);
-export const UNWRAPPED_MAINNET_ETH_REGISTRAR_CONTRACT = buildContractRef(
-  MAINNET,
-  buildAddress("0x57f1887a8BF19b14fC0dF6Fd9B2acc9Af147eA85"),
-);
+  public getManagedSubname = (name: ENSName): ENSName | null => {
```

Comment:
Should this function be called `getDirectSubname`?

Reply:
No. Sometimes the managed subname isn't direct.